### ISSUE TITLE
(pipelines): make bundling of lambdas in asset stage as container images are done

### ISSUE THREAD

**Issue by Hi-Fi** [aws/aws-cdk#21848]
### Describe the feature

Currently Lambdas are bundled along the synth, which causes issues for duration of both synth and also tests that test the templates.

If the definition would be created same way as for container images, synth and tests would be quite a lot lighter, and utilization of parallelization in asset stage would benefit users.

### Use Case

This change would help user by providing
- Speed up synth
- Speed up test runs
- Prevent need of secrets locally (if bundling requires those, e.g. for access to private registries)

### Proposed Solution

Synth should generate same kind of asset definition and copy needed files to cdk.out that's currently done for container based assets.

### Acknowledgements

- [x] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

N/A

### Environment details (OS name and version, etc.)

N/A

**Comment by rix0rrr** [comment]
Yep, I agree. I am unhappy with how bundling works currently, and would much prefer it be a discrete preprocessing step.

There is one downside to that: it would be impossible to have the asset hash be the hash of the *output* artifacts (as opposed to the *input* artifacts).

![image](https://user-images.githubusercontent.com/524162/188167424-d8dd678c-c5c4-41e6-a1e0-990c6ef33d42.png)

That means that any nondeterministic commands like `apt-get` or `curl` in the bundling command (and remember, it can be arbitrary commands run inside a Docker container) cannot be accurately reflected in the asset hash.

**Comment by Hi-Fi** [comment]
Probably duplicate with #19157

**Comment by Hi-Fi** [comment]
Isn't the actual place where this functionality is defined at here?
https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/core/lib/asset-staging.ts#L186-L193

**Comment by Hi-Fi** [comment]
@rix0rrr Not sure if you checked #22325, but isn't that something that's required to do this bundling happen at separate step? As I don't see that it would be reasonable to start creating any pipelines specific hacks.
